### ISSUE TITLE
Unexpected acceptance and lowering of "dependent type" (spurious TypeVar?)

### ISSUE THREAD

**Issue by phipsgabler** [JuliaLang/julia#37774]
Let

```julia
struct Lebesgue{X} end
Lebesgue(X) = Lebesgue{X}()
```

I was very surprised to find the following function to work:

```julia
julia> g(x::X, ::typeof(Lebesgue(eltype(Vector{X})))) where {X<:Real} = ()
g (generic function with 1 method)
```

However, it does not even do what it looks like:

```julia
julia> methods(g)
# 1 method for generic function "g":
[1] g(x::X, ::Lebesgue{Any}) where X<:Real in Main at REPL[70]:1

julia> g(1, Lebesgue(Int))
ERROR: MethodError: no method matching g(::Int64, ::Lebesgue{Int64})
Closest candidates are:
  g(::X, ::Lebesgue{Any}) where X<:Real at REPL[70]:1
Stacktrace:
 [1] top-level scope at REPL[74]:1

julia> g(1, Lebesgue(String))
ERROR: MethodError: no method matching g(::Int64, ::Lebesgue{String})
Closest candidates are:
  g(::X, ::Lebesgue{Any}) where X<:Real at REPL[70]:1
Stacktrace:
 [1] top-level scope at REPL[75]:1
```

On the contraty, something like the following is immediately rejected:

```julia
julia> g2(x::X, ::typeof(Lebesgue(X))) where {X<:Real} = ()
ERROR: UndefVarError: X not defined
Stacktrace:
 [1] (::Type{Lebesgue{X<:Real}})() at ./REPL[17]:1
 [2] Lebesgue(::TypeVar) at ./REPL[18]:1
 [3] top-level scope at REPL[93]:1
```

But `X` in `g` is valid as a free `TypeVar`, since it appears in a type application:

```julia
Meta.@lower g(x::X, ::typeof(Lebesgue(eltype(Vector{X})))) where {X<:Real} = ()
:($(Expr(:thunk, CodeInfo(
    @ none within `top-level scope'
1 ─       $(Expr(:thunk, CodeInfo(
    @ none within `top-level scope'
1 ─     return $(Expr(:method, :g))
)))
│         $(Expr(:method, :g))
│   %3  = Core.TypeVar(:X, Real)
│   %4  = Core.Typeof(g)
│   %5  = Core.apply_type(Vector, %3)
│   %6  = eltype(%5)
│   %7  = Lebesgue(%6)
│   %8  = typeof(%7)
│   %9  = Core.svec(%4, %3, %8)
│   %10 = Core.svec(%3)
│   %11 = Core.svec(%9, %10, $(QuoteNode(:(#= In[3]:4 =#))))
│         $(Expr(:method, :g, :(%11), CodeInfo(quote
    Core.tuple()
    return %1
end)))
└──       return g
))))
```

And then we have a case like this:

```julia
julia> XX = TypeVar(:XX)
XX

julia> Vector{XX}
Array{XX,1}

julia> eltype(Vector{XX})
Any
```

I guess it shouldn't even come that far?

**Comment by yuyichao** [comment]
1. The `typeof` is evaluated on the spot. It is never part of the method signature. What you are observing is simply the evaluation of `typeof(Lebesgue(eltype(Vector{X})))` and `typeof(Lebesgue(X))` when `X` is a `TypeVar` and has nothing to do with method definition (everything you observed happens before that step). This is a useful feature that should not be removed but it is certainly one that must be used with care.

2. The one with `Vector` does not error since there's nothing wrong with constructing types with free type var, those types are simply "incomplete". OTOH, you cannot construct an object with such a type which is what give you the error in the other version.

> I guess it shouldn't even come that far?

In another word, yes, it should get this far and I don't think there's anything to be changed about it.

**Comment by simeonschaub** [comment]
Note that the reason the second one fails is because `Lebesgue(X)` ends up trying to call `Lebesgue{TypeVar(:X)}()`, which still has free `TypeVar`s. It's the same reason the following fails:
```julia
julia> T = Val{TypeVar(:y)}
Val{y}

julia> T()
ERROR: UndefVarError: x not defined
Stacktrace:
 [1] (::Type{Val{y}})() at ./essentials.jl:694
 [2] top-level scope at REPL[6]:1
```
I do think we should be able to add a check in `jl_apply_generic` whether the function is a `DataType` with free `TypeVar`s and throw a better error in these cases, since the current behavior of throwing an `UndefVarError` in a function the user probably didn't even write is pretty confusing.

**Comment by phipsgabler** [comment]
OK, got it, but it's still weird that a bound variable suddenly behaves like a free variable,  in some sense.  This is somewhat contradictory, isn't it?

On the other hand, a more minimal example of the same effect is probably

```
typeof(Lebesgue(eltype(Vector{X}))) where {X}
```

And if I see that, `Lebesgue{Any}` is the only type that makes sense.

**Comment by yuyichao** [comment]
> a bound variable suddenly behaves like a free variable

What do you mean a bound one becomes a free one?

**Comment by phipsgabler** [comment]
An expression `e where {X}`, with `X` is free in `e`, IMHO should either fail because `X` is used in a wrong way (`foo(X) where X`), or describe a the `UnionAll` type implied by it (`Vector{X} where {X}`), making `X` bound.  But instead, it behave exactly like `e` with `X` substituted for `TypeVar(:X)` -- an expression containing a free variable.

**Comment by simeonschaub** [comment]
> `X` is used in a wrong way (`foo(X) where X`)

This is not necessarily "wrong", because `foo` could be defined as `foo(x) = Val{x}`, so this would be perfectly valid.

**Comment by phipsgabler** [comment]
OK, that absolutely makes sense.  Is there anywhere a complete description of how the conversion from `e where {X}` to a `UnionAll` works?

**Comment by simeonschaub** [comment]
You can just look at the output of `Meta.@lower`, but it's also explained quite well here: https://docs.julialang.org/en/v1/devdocs/types/

**Comment by phipsgabler** [comment]
I think I get it now.  `e where {X}` becomes

```julia
let T = TypeVar(:T)
    UnionAll(T, e[X => T])
end
```

So if `e` contains a type application, everything is well, and if `e` contains a function application of `X`, then this gets the `TypeVar` that stands for the `X`, and may do whatever it likes.

The `Any` result of my original example is only an artifact of `eltype`, since

```julia
eltype(::Type{<:AbstractArray{E}}) where {E} = @isdefined(E) ? E : Any
```

Thanks for your patience!

**Comment by JeffBezanson** [comment]
Yes, correct. It would be nice if we could give some sort of better error, but as Yichao said I can't see how to do that without breaking existing useful cases. And most of the bad behavior here is indeed due to the "fallback" `eltype` that allows the eltype of anything to be `Any`. We have tried in the past to get rid of that, but hopefully we can do it for real in 2.0.
